docs/source/isolation.rst

```diff
@@ -0,0 +1,200 @@
+At Yelp, we used the '--memory-swap' parameter to tell Docker to set this value
+to the *same value as the memory parameter*. This prevents a container from swapping at all.
+
+You can see these values by looking at:
+
+* `cat /sys/fs/cgroup/docker/<container-id>/memory.limit_in_bytes`
+* `cat /sys/fs/cgroup/docker/<container-id>/memory.memsw.limit_in_bytes`
+
+In Yelp's setup, these values should be the same.
+
+If the processes in the cgroup reaches the ``memsw.limit_in_bytes`` value ,
+then the kernel will invoke the OOM killer, which in turn will kill off one of
+the processes in the cgroup (often, but not always, this is the biggest
+contributor to the memory usage). If this is the only process running in the
+Docker container, then the container will die. The mesos framework which
+launched the task may or may not decide to try and start the same task
+elsewhere.
+
+CPUs
+""""
+
+CPU enforcement is implemented slightly differently. Many people expect the
+value defined in the ``cpus`` field in a service's soa-configs to map to a
+number of cores that are reserved for a task. However, isolating CPU time like
+this can be particularly wasteful; unless a task spends 100% of it's time on
+CPU (and thus has *no* I/O), then there is no need to prevent other tasks from
+running on the spare CPU time available.
+
+Instead, the CPU value is used to give tasks a relative priority. This priority
+is used by the Linux Scheduler decide the order in which to run waiting
+threads.
+
+Some notes on this:
+
+  - As mentioned, these values are relative. We enforce no scale on these
+    numbers. So if two tasks are competing for resources, one has set ``cpus``
+    to 0.5 and the other to 4.5, then it the first task will receive 10% of time
+    and the second 90%. If, however, the second task was replaced with another
+    with a requirement of 0.5, leaving two tasks with each ``cpu`` set to 0.5,
+    then each task will have 50% of time.
+  - The amount of time tasks get is proportional to the number of tasks on the
+    host. If there are 3 tasks on the host, with ``cpu`` values 10,5,5 then the
+    time will be split 50%, 25%, 25%. However, if a fourth task is run, with
+    ``cpu`` set to 1000, then that time becomes 33%, 16.5%, 16.5%, 33%.
+  - Any 'spare' cpu cycles are redistributed by the CPU, so if a task does
+    *not* use it's 'share', then other cgroups will be allocated this spare CPU
+    time.
+  - All threads inside a cgroup are considered when the scheduler decides the
+    fair share of time. That is, if your container launches multiple tasks,
+    then the share is split across all these tasks. If the tasks were to run in
+    their own cgroup, then the time spent on CPU by one task would not count
+    against the share available for another. The result of this may be that
+    a higher number of 'skinny' containers may be preferable to 'fat' containers.
+
+This is different from how Mesos and Marathon use the CPU value when evaluating
```

Comment:
This is a yelp-specific thing, please note that.

Mesos/marathon don't care what you advertise. I still say we should report ECU units or something normalized.